[Feature] vLLM ARM Enablement for AARCH64 CPUs

sanketkaleoss opened the pull request:
**Description**
This PR enables support of vLLM for AARCH64 architecture. Motivated by the requirements from (#176, #5741, etc), I implemented this PR which enables the ARM path for CPU inference.  

**ARM Compatibility:** Modified the build scripts and configuration files to ensure compatibility with ARM processors. It currently supports float32, fp16 and bfloat16 datatypes.

**Motivation**
Enabling vLLM on ARM architecture broadens its usability, allowing it to run on a wider range of devices, including those with ARM processors. This enhancement is crucial for expanding the reach and applicability of vLLM in various use cases.

**Checklist**
- Code changes have been tested on ARM devices (Graviton3).

**Modifications**
1. Added NEON intrinsics for enabling vLLM on ARM
1. Updated requirements-cpu.txt
1. Added ARM path in cpu_extension.cmake

**Performance Results**

Model : facebook/opt-125m
Datatype : float32

![image](https://github.com/user-attachments/assets/e41a83b3-7864-4b4a-9f2b-cc12de291b52)

Commit:
Enabled path for ARM machines

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

Commit:
Enabled path for ARM machines

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

sanketkaleoss committed:
fixed typos

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

sanketkaleoss commented:
@youkaichao Please have a look at this PR. I've added the new path for vLLM to work on ARM CPUs.

sanketkaleoss commented:
@youkaichao @ChipKerchner @mgoin @pathorn @manojnkumar I have added a new feature that allows vLLM to run on ARM CPU backend. I have tested on AWS Graviton 3E.

mgoin commented on cmake/cpu_extension.cmake:
```diff
@@ -68,18 +69,28 @@ if (AVX512_FOUND AND NOT AVX512_DISABLED)
+    message(STATUS "ARMv8 architecture detected")
+    list(APPEND CXX_COMPILE_FLAGS
+        "-mcpu=native"
+        "-march=armv8.6-a"
```
Why was this specific arch chosen?

sanketkaleoss replied:
armv8.6-a is the first architecture that supports advance SIMD instructions, bf16 support and SVE support. That's why I chose this specific arch.

mgoin replied:
This is fine for now, but for instance I think Apple's M1 CPU uses ARMv8.4-A so we should consider supporting older versions

sanketkaleoss replied:
I see, noted.

tlrmchlsmth replied:
[According to wikichip](https://en.wikichip.org/wiki/annapurna_labs/graviton/graviton3), Graviton3 is also ARMv8.4A -- is the Graviton3E different? I haven't found any documentation on it

sanketkaleoss replied:
@tlrmchlsmth  You're right, graviton 3 has armv8.4. The code needs 8.6 only for bf16 dependencies. But if I just use "-mcpu=native" or "-march=armv8.4-a+bf16" the code works fine on Graviton3 instances. What do you suggest here?

tlrmchlsmth replied:
If `-march=armv8.4-a+bf16` works, then I suggest using that for this PR.

In general, I think it's best to go with a fairly minimal ISA version, and then explicitly specify what ISA extensions we want to use, especially because ARM has _many_ extensions that are optional for several ISA versions before they become mandatory

tlrmchlsmth replied:
BTW, the M1 Mac (and I think M2) does not support BF16 NEON so unfortunately this wouldn't help there.

sanketkaleoss replied:
I see, thanks for the information. What do you suggest here, to add macos support for future PR or modify into this PR?

sanketkaleoss replied:
Hi @mgoin , @tlrmchlsmth I've added the dockerfile and documentation. Changed the arm arch to native. Please have a look.

mgoin reviewed (commented):
Nice work @sanketkaleoss! This seems reasonable to me as a base implementation. It seems only compute in fp32 is supported, is that right?
It would be good to update the cpu installation documentation with how to build and to also add a new Dockerfile.arm

sanketkaleoss commented:
> Nice work @sanketkaleoss! This seems reasonable to me as a base implementation. It seems only compute in fp32 is supported, is that right? It would be good to update the cpu installation documentation with how to build and to also add a new Dockerfile.arm

Thanks @mgoin for the review. It supports fp32 and bf16 as of now. Sure, I'll work on updating the documentation and adding Dockerfile too.

youkaichao commented:
this is great, I will hand it over to @mgoin for detailed review.

it would be better if we can support macos m chips as well, they are also arm chips.

sanketkaleoss commented:
> this is great, I will hand it over to @mgoin for detailed review.
> 
> it would be better if we can support macos m chips as well, they are also arm chips.

Thanks @youkaichao . I'll add that to future work.

tlrmchlsmth commented on csrc/cpu/cpu_types_arm.hpp:
```diff
@@ -0,0 +1,457 @@
+
+namespace vec_op {
+
+// FIXME: FP16 is not fully supported in Torch-CPU
```
@sanketkaleoss do you have plans to support FP16 in a future PR? I see that it's partially implemented. 

Do you know what the problem with FP16 in Torch-CPU is?

sanketkaleoss replied:
I'm not sure what's the problem there. I can add the support in future PR if torch-CPU supports FP16 in future.

Commit:
Added Dockerfile for ARM architecture support

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

caijijuhe commented:
Hello, I followed your Dockerfile.arm encountered during the build process。
ERROR: ld.so: object '/usr/lib/aarch64-linux-gnu/libtcmalloc_minimal.so.4' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
Can you share the built image

Commit:
Added Documenataion for ARM architecture support and updated architecture to native

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

Commit:
Updated documentation

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

tlrmchlsmth commented on cmake/cpu_extension.cmake:
```diff
@@ -68,18 +69,28 @@ if (AVX512_FOUND AND NOT AVX512_DISABLED)
         "-mtune=native")
+
+elseif (ASIMD_FOUND)
+    message(STATUS "ARMv8 architecture detected")
```
Since it could be ARMv9:

```suggestion
    message(STATUS "ARMv8 or later architecture detected")
```
or maybe explicitly call out NEON instead

sanketkaleoss replied:
Okay, sure

tlrmchlsmth commented on cmake/cpu_extension.cmake:
```diff
@@ -68,18 +69,28 @@ if (AVX512_FOUND AND NOT AVX512_DISABLED)
+elseif (ASIMD_FOUND)
+    message(STATUS "ARMv8 architecture detected")
+    list(APPEND CXX_COMPILE_FLAGS
+        "-mcpu=native"
+        "-mtune=native"
```
Native is OK for local development, however the binaries won't be portable at all.

I actually thought you were on the right track before this change. We really ought to explicitly specify a fairly minimal base ARM architecture (maybe ARMv8.4) and then explicitly the set of extensions that we need to build (BF16, and maybe FP16 and DotProd?)

sanketkaleoss replied:
I see, I'll try to set it as armv8.4 as a base implementation. Then, depending on the flags it would run bf16 or not.

tlrmchlsmth replied:
Could you try using `"-march=armv8.4-a+bf16+dotprod+fp16"`?

ddynwzh1992 replied:
Just for your guys reference, on arm64 -mcpu= acts as both specifying the appropriate architecture and tuning and it's generally better to use that vs -march if you're building for a specific CPU, you can find more details from here if you are running any test on Graviton: https://github.com/aws/aws-graviton-getting-started/blob/main/c-c++.md

sanketkaleoss replied:
> Could you try using `"-march=armv8.4-a+bf16+dotprod+fp16"`?

Yes, I tried it and it works. It even works with `"-march=armv8.2-a+bf16+dotprod+fp16"` making it compatible with Graviton2.

sanketkaleoss replied:
@tlrmchlsmth I'll fix `"-march=armv8.2-a+bf16+dotprod+fp16"` for this PR, and make it backward compatible with mac chips if the issue is raised for that in future PR. What do you suggest?

Commit:
Added compatibility with macos by creating separate fp32 and bf16 paths

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

sanketkaleoss commented:
@mgoin @tlrmchlsmth I have created separate paths for FP32 and BF16 datatypes. This code now works on any arm machine from `armv8.2-a` onwards. Even if BF16 support is not present, it will choose FP32 path. Please have a look.

Commit:
Fixed some missing conditions

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

Commit:
fixed dco test

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

Commit:
Rebased and resolved merge conflicts

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

Commit:
Updated documentation

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

Commit:
Added FP16 compatibility and updated torch version

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

Commit:
Resolved merge conflicts

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

sanketkaleoss commented:
Hi @mgoin @tlrmchlsmth @ddynwzh1992 , I've implemented the suggested changes and now ARM CPU backend path supports FP32, FP16 and BF16 datatypes. Support for MAC devices has been added. I've tested on ` -march = armv8.2-a` onwards and it runs fine even without the added extensions. Please review the changes.

Commit:
Resolve failed formatting checks

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

ShawnD200 commented:
> Even if BF16 support is not present, it will choose FP32 path.

Maybe I missed something, how will it choose FP32 path if bp16 not supported? Thank you.

sanketkaleoss commented:
> > Even if BF16 support is not present, it will choose FP32 path.
> 
> Maybe I missed something, how will it choose FP32 path if bp16 not supported? Thank you.

Earlier it used to give compilation error even if we just want to use FP32 and not BF16, as there were no separate paths for FP32 and Bf16 in x86 backend case. Now, it will compile successfully even if the system doesn't have BF16 extension. User can run by selecting FP32 dtype in that case.

sanketkaleoss commented:
@mgoin @tlrmchlsmth @WoosukKwon @youkaichao @DarkLight1337 Your attention is required on this PR. Thanks.

ShawnD200 commented on csrc/cpu/cpu_types_arm.hpp:
```diff
@@ -0,0 +1,520 @@
+        if (remainder > 0) {
+            float16x8_t temp = reg.val[full_blocks];
+            for (int i = 0; i < remainder; ++i) {
+                reinterpret_cast<__fp16*>(ptr)[full_blocks * 8 + i] = vgetq_lane_f16(temp, i);
```
Hello, are you sure this line can compile? vgetq_lane_f16 must have compile-time constant index.

ShawnD200 commented:
> > > Even if BF16 support is not present, it will choose FP32 path.
> > 
> > 
> > Maybe I missed something, how will it choose FP32 path if bp16 not supported? Thank you.
> 
> Earlier it used to give compilation error even if we just want to use FP32 and not BF16, as there were no separate paths for FP32 and Bf16 in x86 backend case. Now, it will compile successfully even if the system doesn't have BF16 extension. User can run by selecting FP32 dtype in that case.

I thought you meant BF16 is supported without its hardware feature, then macbook won't have it as of now, yes?

ShawnD200 commented on csrc/cpu/cpu_types_arm.hpp:
```diff
@@ -0,0 +1,520 @@
+        if (remainder > 0) {
+            float16x8_t temp = reg.val[full_blocks];
+            for (int i = 0; i < remainder; ++i) {
+                reinterpret_cast<__fp16*>(ptr)[full_blocks * 8 + i] = vgetq_lane_f16(temp, i);
```
Hello, are you sure this line can compile? The index must be a compile-time constant, no?

sanketkaleoss replied:
Thanks for the review. It compiles on my system as of now.

ShawnD200 replied:
Thanks for the reply. I found that with gcc v12.3 (Ubuntu), optimization O1 and above does not require const (while O0 does), with clang v15.0 (macOS), I haven't been able to compile it.

sanketkaleoss replied:
> Thanks for the reply. I found that with gcc v12.3 (Ubuntu), optimization O1 and above does not require const (while O0 does), with clang v15.0 (macOS), I haven't been able to compile it.

I see, thanks for the information.

mgoin commented on csrc/cpu/cpu_types_arm.hpp:
```diff
@@ -0,0 +1,520 @@
+      reg.val[3] = vcvt_f32_f16(vget_high_f16(v.reg.val[1]));
+  };
+
+  // #ifdef BF16_SUPPORT
+  // explicit FP32Vec16(const BF16Vec8 &v) : FP32Vec16(FP32Vec8(v)) {}
+  // #endif
```
Is this not needed?

sanketkaleoss replied:
It is defined elsewhere, I'll remove these comments.

mgoin reviewed (commented):
This essentially looks good to me, left a few comments. Please resolve the merge conflict and I can set the ready tag for CI to run.

I was able to test and run this on my M1 Macbook - nice work!

Server:
```
docker build -f Dockerfile.arm -t vllm-cpu-env --shm-size=4g . 
docker run -v ~/.cache/huggingface:/root/.cache/huggingface -p 8000:8000 --ipc=host vllm-cpu-env --model Qwen/Qwen2.5-0.5B-Instruct --dtype float16
```

Client:
```
curl http://0.0.0.0:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer DUMMY" \
  -d '{
    "model": "Qwen/Qwen2.5-0.5B-Instruct",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "What is the best pie?"
      }
    ]
  }'
```

Output:
```
{"id":"chatcmpl-210cbdc345ce48d9988a9aced2f83300","object":"chat.completion","created":1732058836,"model":"Qwen/Qwen2.5-0.5B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"As an AI language model, I don't have personal preferences or tastes, but I can provide information about different types of pies based on popular opinion.\n\nThe best pie for everyone depends on personal taste, preferences, and dietary needs. Here are some popular types of pies:\n\n1. Pumpkin pie: It is a classic pie that uses pumpkins as a main ingredient. It is often served with crème fraîche, whipped cream, or chocolate ganache.\n\n2. Apple pie: Apple pie is a sweet and fruity pie that is popular in the autumn season. It uses apples as a main ingredient and is often served with whipped cream or chocolate ganache.\n\n3. Blueberry pie: Blueberry pie is a classic pie that uses blueberries as a main ingredient. It is often served with whipped cream or chocolate ganache.\n\n4. Chocolate pie: Chocolate pie is a sweet and rich pie that is popular in the winter season. It uses chocolate as a main ingredient and is often served with whipped cream or chocolate ganache.\n\n5. Lemon pie: Lemon pie is a fresh and zesty pie that is popular in the summer season. It uses lemons as a main ingredient and is often served with whipped cream or chocolate ganache.\n\nUltimately, the best pie for everyone depends on their personal preferences and dietary needs.","tool_calls":[]},"logprobs":null,"finish_reason":"stop","stop_reason":null}],"usage":{"prompt_tokens":25,"total_tokens":290,"completion_tokens":265,"prompt_tokens_details":null},"prompt_logprobs":null}
```

sanketkaleoss commented on cmake/cpu_extension.cmake:
```diff
@@ -53,6 +58,8 @@ find_isa(${CPUINFO} "avx2" AVX2_FOUND)
 find_isa(${CPUINFO} "POWER10" POWER10_FOUND)
 find_isa(${CPUINFO} "POWER9" POWER9_FOUND)
+find_isa(${CPUINFO} "asimd" ASIMD_FOUND) # Check for ARM NEON support
+find_isa(${CPUINFO} "bf16" BF16_FOUND) # Check for BF16 support
```
Sure, I'll do it

sanketkaleoss commented on cmake/cpu_extension.cmake:
```diff
@@ -16,9 +16,14 @@ include_directories("${CMAKE_SOURCE_DIR}/csrc")
 # Check the compile flags
 #
+
+if (NOT CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
+    list(APPEND CXX_COMPILE_FLAGS
+        "-mf16c"
+    )
+endif()
```
You're right, it is specifiv to x86. I'll modify it.

Commit:
Changed flag name and modified compile flag declaration

Signed-off-by: Sanket Kale <sanketk.kale@fujitsu.com>

sanketkaleoss commented:
> This essentially looks good to me, left a few comments. Please resolve the merge conflict and I can set the ready tag for CI to run.
> 
> I was able to test and run this on my M1 Macbook - nice work!
> 
> Server:
> 
> ```
> docker build -f Dockerfile.arm -t vllm-cpu-env --shm-size=4g . 
> docker run -v ~/.cache/huggingface:/root/.cache/huggingface -p 8000:8000 --ipc=host vllm-cpu-env --model Qwen/Qwen2.5-0.5B-Instruct --dtype float16
> ```
> 
> Client:
> 
> ```
> curl http://0.0.0.0:8000/v1/chat/completions \
>   -H "Content-Type: application/json" \
>   -H "Authorization: Bearer DUMMY" \
>   -d '{
>     "model": "Qwen/Qwen2.5-0.5B-Instruct",
>     "messages": [
>       {
>         "role": "system",
>         "content": "You are a helpful assistant."
>       },
>       {
>         "role": "user",
>         "content": "What is the best pie?"
>       }
>     ]
>   }'
> ```
> 
> Output:
> 
> ```
> {"id":"chatcmpl-210cbdc345ce48d9988a9aced2f83300","object":"chat.completion","created":1732058836,"model":"Qwen/Qwen2.5-0.5B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"As an AI language model, I don't have personal preferences or tastes, but I can provide information about different types of pies based on popular opinion.\n\nThe best pie for everyone depends on personal taste, preferences, and dietary needs. Here are some popular types of pies:\n\n1. Pumpkin pie: It is a classic pie that uses pumpkins as a main ingredient. It is often served with crème fraîche, whipped cream, or chocolate ganache.\n\n2. Apple pie: Apple pie is a sweet and fruity pie that is popular in the autumn season. It uses apples as a main ingredient and is often served with whipped cream or chocolate ganache.\n\n3. Blueberry pie: Blueberry pie is a classic pie that uses blueberries as a main ingredient. It is often served with whipped cream or chocolate ganache.\n\n4. Chocolate pie: Chocolate pie is a sweet and rich pie that is popular in the winter season. It uses chocolate as a main ingredient and is often served with whipped cream or chocolate ganache.\n\n5. Lemon pie: Lemon pie is a fresh and zesty pie that is popular in the summer season. It uses lemons as a main ingredient and is often served with whipped cream or chocolate ganache.\n\nUltimately, the best pie for everyone depends on their personal preferences and dietary needs.","tool_calls":[]},"logprobs":null,"finish_reason":"stop","stop_reason":null}],"usage":{"prompt_tokens":25,"total_tokens":290,"completion_tokens":265,"prompt_tokens_details":null},"prompt_logprobs":null}
> ```

Thanks for the detailed review @mgoin ! I have rebased and implemented the suggested changes. Please proceed with the CI checks.

sanketkaleoss committed:
Merge branch 'main' into vllm_arm_enablement

mgoin committed:
Merge branch 'main' into vllm_arm_enablement

mgoin reviewed (approved):
Sorry for losing track, merged with main to see if green

sanketkaleoss commented:
> Sorry for losing track, merged with main to see if green

No worries, looking forward to the future work on this PR.

animalnots commented:
@sanketkaleoss Hi, thank you for the PR. Does additional work needs to be done to support `Qwen/Qwen2-VL-2B-Instruct` ?

sidharthrajaram commented:
@animalnots Did you figure that out? There's an issue related to it: https://github.com/vllm-project/vllm/issues/11154

animalnots commented:
> @animalnots Did you figure that out? There's an issue related to it: https://github.com/vllm-project/vllm/issues/11154

I ended up using this 

https://github.com/satvikahuja/Easy-qwen2vlm2b-4macbook/issues/1

Vllm never worked and instead would output "!!!!!!!!!..."

sanketkaleoss commented:
> @sanketkaleoss Hi, thank you for the PR. Does additional work needs to be done to support `Qwen/Qwen2-VL-2B-Instruct` ?

Hi, sorry for the late reply. Does the model work with x86 CPU path?
